[14.0][IMP] cetmix_tower_git-link/copy git project on file command

Kurama2d opened the pull request:
Enable automatic git project relation when using file template
in flight plans.

This removes the need to manually link files like addons.yaml
to a git project after server creation.

Task: 4759

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added the ability to link a Git project to a flight plan line, with an option to create a copy of the Git project.
  * Enhanced the server creation process to support associating servers and files with Git projects based on flight plan lines.
  * Updated the user interface to display Git project fields on the relevant plan line form when applicable.

* **Bug Fixes**
  * None.

* **Tests**
  * Introduced tests to verify server creation with Git project linkage and file generation from flight plan lines.

* **Chores**
  * Updated internal module and test initializations to support new features.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Kurama2d commented on cetmix_tower_git/README.rst:
```diff
@@ -13,7 +13,7 @@ Cetmix Tower Git
     :target: https://odoo-community.org/page/development-status
     :alt: Production/Stable
-.. |badge2| image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
     :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
     :alt: License: AGPL-3
```
It's automatically created line

Kurama2d commented on cetmix_tower_git/tests/test_plan_line.py:
```diff
@@ -0,0 +1,129 @@
+            }
+        )
+
+        # Create server from template
+        with patch(
+            "odoo.addons.cetmix_tower_server.models.cx_tower_server.CxTowerServer.upload_file",
+            return_value="ok",
+        ), patch(
+            "odoo.addons.cetmix_tower_server.models.cx_tower_server.CxTowerServer.run_command",
+            return_value={"status": 0, "response": "Success", "error": None},
+        ):
```
I thing this double patch is okay.

Kurama2d commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,103 @@
+                return res
+            git_project_id = git_project_id.copy()
+        else:
+            existing_rel = rel_obj.search(
```
it's actually existing relation, not conflictiong one

Kurama2d commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,103 @@
+
+        git_project_id = self.git_project_id
+        if not git_project_id:
+            return res
```
we have ensure_one() in original _run which we call with super

Kurama2d commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,103 @@
+                ("file_type", "=", "text"),
+            ],
+            limit=1,
+        )
```
Fixed it

Kurama2d commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,102 @@
+            )
+
+            if existing_rel:
+                return res
```
Fixed.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_plan_line.py:
You should override the [command runner](https://github.com/cetmix/cetmix-tower/blob/46eb0c24983318fe7edca11342410dc7f521ab47/cetmix_tower_server/models/cx_tower_server.py#L1165) function instead. This is what those functions are made for.

Kurama2d replied:
Thanks. method is overriden and logic is changed

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,131 @@
+            [("server_id", "=", server.id), ("name", "=", "repos.yaml")]
+        )
+
+        self.assertTrue(file, "Git config file should have been created")
```
Changed

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,131 @@
+            ]
+        )
+
+        self.assertTrue(git_project_rel, "Git project relation should exist")
```
Changed

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,131 @@
+            git_project_rel.project_format,
+            git_project._default_project_format(),
+            "Project format should match the default format",
+        )
```
I prefer to create and use data in test directly

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,123 @@
+    def setUp(self):
+        super().setUp()
+
+    def test_server_creation_running_flight_plan(self):
```
Changed

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,123 @@
+
+        self.assertTrue(git_project_rel, "Git project relation should exist")
+        self.assertEqual(
+            git_project_rel.file_id,
+            file,
+            "The related file should be the git config file",
+        )
+        self.assertEqual(
+            git_project_rel.git_project_id,
+            git_project,
+            "The related git project should match the one in the flight plan",
```
I prefer recordset compare, it works good

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,123 @@
+
+
+class TestSever(CommonTest):
+    """Test setting git project to server from plan line."""
```
Changed

Kurama2d commented on cetmix_tower_git/tests/test_server.py:
```diff
@@ -0,0 +1,123 @@
+            ]
+        )
+
+        self.assertTrue(server, "Server should have been created")
```
Changed

Kurama2d commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,21 @@
+
+    _inherit = "cx.tower.plan.line"
+
+    git_project_id = fields.Many2one(
+        comodel_name="cx.tower.git.project",
+        string="Git Project",
+        help="Select a git project to be linked to the file and server.",
+    )
+    is_make_copy = fields.Boolean(
+        string="Make a Copy",
+        help="Create a copy of the Git Project instead of linking "
+        "the file to the existing one.",
+    )
```
It's not said in the task that we need to limit this fields usage by group, so it's not needed here

Kurama2d commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -193,6 +193,7 @@ def _run(self, server, plan_log_record, **kwargs):
         log_vals = kwargs.get("log", {})
         log_vals.update({"plan_log_id": plan_log_record.id})
         kwargs.update({"log": log_vals})
+        kwargs.update({"plan_line_id": self})
 
```
From this moment I need relevant data and can't keep previous one to be used later

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1162,6 +1162,16 @@ def _command_runner(
+    def _command_runner_file_using_template_create_file(
+        self, file_template_id, server_dir, **kwargs
+    ):
+        """Create a file using the specified file template on the server."""
```
Please add the complete docstring. It also must explain why do we need this function, otherwise it's not clear from the code.

Kurama2d replied:
Extended the docstring.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_server.py:
```diff
@@ -36,3 +36,34 @@ def _get_fields_for_yaml(self):
+        """Override to create git project relation
+        when creating a file using a template.
+        """
+        file_id = super()._command_runner_file_using_template_create_file(
```
the `create_file` function of the `cx.tower.file.template` model returns a [file](https://github.com/cetmix/cetmix-tower/blob/ac9775b6a0723d3f5575c036ab6f165ba6429755/cetmix_tower_server/models/cx_tower_file_template.py#L146), not a file ID.

Kurama2d replied:
Checked oca naming rules and you're right, we don't use _id for variable with record variable anymore.

Kurama2d commented on cetmix_tower_git/models/cx_tower_server.py:
```diff
@@ -36,3 +36,34 @@ def _get_fields_for_yaml(self):
+                    "project_format": git_project._default_project_format(),
+                }
+            )
+        return file
```
It's extra here

Kurama2d commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1162,6 +1162,24 @@ def _command_runner(
+            server=self,
+            server_dir=server_dir,
+            raise_if_exists=True,
+        )
```
Changed

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_plan_line.py:
```diff
@@ -193,6 +193,7 @@ def _run(self, server, plan_log_record, **kwargs):
         log_vals = kwargs.get("log", {})
         log_vals.update({"plan_log_id": plan_log_record.id})
         kwargs.update({"log": log_vals})
+        kwargs.update({"plan_line": self})
```
No need to add the plan line here. Because it can be fetched from the [plan log](https://github.com/cetmix/cetmix-tower/blob/0eb40a1081bf2cb307e613a1284a98593cb9b880/cetmix_tower_server/models/cx_tower_command_log.py#L81) currently executed [line](https://github.com/cetmix/cetmix-tower/blob/0eb40a1081bf2cb307e613a1284a98593cb9b880/cetmix_tower_server/models/cx_tower_plan_log.py#L51).

Kurama2d replied:
I saw that we can get the plan_line through the plan_log, either by using the ID from kwargs in the _command_runner_file_using_template_create_file method directly, or through self.plan_log_ids. But since plan_log_ids is a One2many, there can be more than one log.

Passing the plan_line directly seems safer, because other values and calculations depend on it. It also helps keep the data if something goes wrong with the plan_log.

ivs-cetmix replied:
Nope, I don't mean `self.plan_log_ids`.  Please check my reply in the `cx.tower.server` below.

Kurama2d replied:
Maybe it's not clear, but I gave example with `self.plan_log_ids` to show that we can't use it, it wasn't my suggestion.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,21 @@
+# Copyright (C) 2022 Cetmix OÜ
```
2025

Kurama2d replied:
It's not clear why 2025. There are some points:

- In this [commit ](https://github.com/cetmix/cetmix-tower/blob/bfc783933d5df2b2e086e8fe0f5343afde1add74/cetmix_tower_git/models/cx_tower_file_template.py) file was added in 2025, but copyright year is 2024.

- In documentation it's said that copyright year is the year of creation module (nothing about models) and I took copytight year from the original `cx.tower.plan.line` model which is 2022.

So, if in this file should be 2025, maybe it sould be 2025 in all other files which was created in 2025 in this module and I need to fix it?

Fixed it here.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1197,10 +1221,8 @@ def _command_runner_file_using_template(
-                server=self,
-                server_dir=server_dir,
-                raise_if_exists=True,
+            file = self._command_runner_file_using_template_create_file(
```
```python
plan_line = log_record.plan_log_id.plan_line_executed_id
file = _command_runner_file_using_template_create_file(file_template_id=file_template_id, server_dir=server_dir, plan_line=plan_line)
...
```

Kurama2d replied:
Replaced to this logic, but direct passing with kwargs is still a better option.

ivs-cetmix replied:
I understand your concerns. `kwargs` may be a better solution, however we need to keep the code as clean and simple as possible. And adding a new kwargs element is always error prone.

ivs-cetmix reviewed (approved):
Code review LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

GabbasovDinar reviewed (changes requested):
no tested, LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix commented:
/ocabot rebase

CetmixGitDrone commented:
Congratulations, PR rebased to [14.0](https://github.com/cetmix/cetmix-tower/commits/14.0).

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_plan_line.py:
```diff
@@ -0,0 +1,21 @@
+        string="Git Project",
+        help="Select a git project to be linked to the file and server.",
+    )
+    is_make_copy = fields.Boolean(
```
Let's rename this field to `git_project_make_copy` to be more relevant.

Kurama2d replied:
Renamed.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_plan_line.py:
Please inherit [this function](https://github.com/cetmix/cetmix-tower/blob/391b7b8749f81886e75f9f62fbfcd77050f46e13/cetmix_tower_yaml/models/cx_tower_plan_line.py#L10) and add new fields in the list of the exported ones.

Kurama2d replied:
Inherited.

ivs-cetmix reviewed (approved):
Code review LGTM

Kurama2d commented on cetmix_tower_server/models/cx_tower_server.py:
```diff
@@ -1193,10 +1220,12 @@ def _command_runner_file_using_template(
                 if_file_exists=log_record.command_id.if_file_exists,
                 server_dir=server_dir,
+                plan_line=plan_line,
             )
```
fixed it

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

Kurama2d commented on cetmix_tower_git/readme/newsfragments/4759.feature:
```diff
@@ -0,0 +1 @@
+Link or copy a git project when uploading the linked file using command
```
Using command is better here

ivs-cetmix reviewed (approved):
/ocabot merge minor

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [14.0-ocabot-merge-pr-315-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/14.0-ocabot-merge-pr-315-by-ivs-cetmix-bump-minor), awaiting test results.

ivs-cetmix commented on cetmix_tower_git/models/cx_tower_plan_line.py:
Please add the new fields to the list of the yaml exported ones. Same as [here](https://github.com/cetmix/cetmix-tower/blob/be76d1c8a3b7935b7790153b4006853038d89747/cetmix_tower_git/models/cx_tower_server.py#L33)

Kurama2d replied:
Added yaml method extention.

ivs-cetmix commented on cetmix_tower_server/models/cx_tower_server.py:
Please add a newsfragment to this module too:
`4759.feature`
`Improve the extendability of the file upload command`

Kurama2d replied:
Added this file too.

Kurama2d commented on cetmix_tower_server/readme/newsfragments/4759.feature:
```diff
@@ -0,0 +1 @@
+Improve the extendability of the file upload command.
```
Generic description is fine here.

ivs-cetmix committed:
[IMP] cetmix_tower_server File upload command

Improve the extendability of the file upload command.

Task 4759

ivs-cetmix committed:
[IMP] cetmix_tower_git Linked file to project

Link or copy a git project when uploading the linked file using command.

Task 4759

ivs-cetmix commented:
/ocabot merge minor

CetmixGitDrone commented:
Hey, thanks for contributing! Proceeding to merge this for you.
Prepared branch [14.0-ocabot-merge-pr-315-by-ivs-cetmix-bump-minor](https://github.com/cetmix/cetmix-tower/commits/14.0-ocabot-merge-pr-315-by-ivs-cetmix-bump-minor), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 304f8d48e8289cee04fa61c1796e8a79cb5d334e. Thanks a lot for contributing to cetmix. ❤️
